apps/tangle-dapp/data/blueprints/utils/toPrimitiveBlueprint.ts

```diff
@@ -0,0 +1,309 @@
+  if (gadget.type === 'Container') {
+    return toPrimitiveContainerGadget(gadget.asContainer);
+  }
+
+  if (gadget.type === 'Native') {
+    return toPrimitiveNativeGadget(gadget.asNative);
+  }
+
+  if (gadget.type === 'Wasm') {
+    return toPrimitiveWasmGadget(gadget.asWasm);
+  }
+
+  throw new Error('Unknown Gadget type');
+}
+
+export function toPrimitiveWasmGadget({ runtime, sources }: WasmGadget) {
+  return {
+    runtime: toPrimitiveWasmRuntime(runtime),
+    sources: sources.map(toPrimitiveGadgetSource),
+  } as const;
+}
+
+export function toPrimitiveWasmRuntime(runtime: WasmRuntime) {
+  return runtime.type;
+}
+
+export function toPrimitiveNativeGadget({ sources }: NativeGadget) {
+  return {
+    sources: sources.map(toPrimitiveGadgetSource),
+  } as const;
+}
+
+export function toPrimitiveContainerGadget({ sources }: ContainerGadget) {
+  return {
+    sources: sources.map(toPrimitiveGadgetSource),
+  } as const;
+}
+
+export function toPrimitiveGadgetSource(source: GadgetSource) {
+  return {
+    fetcher: toPrimitiveGadgetSourceFetcher(source.fetcher),
+  } as const;
+}
+
+export function toPrimitiveGadgetSourceFetcher(fetcher: GadgetSourceFetcher) {
+  if (fetcher.type === 'ContainerImage')
+    return {
+      ContainerImage: toPrimitiveContainerImage(fetcher.asContainerImage),
+    } as const;
+
+  if (fetcher.type === 'Github')
+    return {
+      Github: toPrimitiveGithubFetcher(fetcher.asGithub),
+    } as const;
+
+  if (fetcher.type === 'Ipfs')
+    return {
+      Ipfs: u8aToString(fetcher.asIpfs),
+    } as const;
+
+  if (fetcher.type === 'Testing')
+    return {
+      Testing: toPrimitiveTestingFetcher(fetcher.asTesting),
+    } as const;
+
+  throw new Error('Unknown GadgetSourceFetcher type');
+}
+
+export function toPrimitiveContainerImage({
+  registry_,
+  image,
+  tag,
+}: ImageRegistryFetcher) {
+  return {
+    registry_: u8aToString(registry_),
+    image: u8aToString(image),
+    tag: u8aToString(tag),
+  } as const;
+}
+
+export function toPrimitiveGithubFetcher({
+  owner,
+  repo,
+  tag,
+  binaries,
+}: GithubFetcher) {
+  return {
+    owner: u8aToString(owner),
+    repo: u8aToString(repo),
+    tag: u8aToString(tag),
+    binaries: binaries.map(toPrimitiveGadgetBinary),
+  } as const;
+}
+
+export function toPrimitiveTestingFetcher({
+  cargoBin,
+  cargoPackage,
+  basePath,
+}: TestFetcher) {
+  return {
+    cargoPackage: u8aToString(cargoPackage),
+    cargoBin: u8aToString(cargoBin),
+    basePath: u8aToString(basePath),
+  } as const;
+}
+
+export function toPrimitiveGadgetBinary({
+  arch,
+  os,
+  name,
+  sha256,
+}: GadgetBinary) {
+  return {
+    arch: toPrimitiveArchitecture(arch),
+    os: toPrimitiveOperatingSystem(os),
+    name: u8aToString(name),
+    sha256,
+  } as const;
+}
+
+export function toPrimitiveArchitecture(arch: Architecture) {
+  return arch.type;
+}
+
+export function toPrimitiveOperatingSystem(os: OperatingSystem) {
+  return os.type;
+}
```

Comment:
Might be simpler if we get rid of these functions and use the properties directly on the consumers